numpy/core/src/multiarray/compiled_base.c

```diff
@@ -80,6 +80,374 @@ minmax(const npy_intp *data, npy_intp data_len, npy_intp *mn, npy_intp *mx)
+                if (ix == bins || (ix != 0 && tx < e[ix])) {
+                    ix--;
+                }
+                else if ((ix + 1) != bins && tx >= e[ix + 1]) {
+                    ix++;
+                }
+                dhist[ix] += w[i];
+            }
+        }
+        NPY_END_ALLOW_THREADS;
+
+        Py_DECREF(arr_w);
+    }
+
+    Py_DECREF(arr_edges);
+    Py_DECREF(arr_x);
+    return (PyObject *)hist;
+
+fail:
+    Py_XDECREF(hist);
+    Py_XDECREF(arr_w);
+    Py_XDECREF(arr_edges);
+    Py_XDECREF(arr_x);
+    return NULL;
+}
+
+NPY_NO_EXPORT PyObject *
+arr_histogramdd_uniform(PyObject *NPY_UNUSED(self), PyObject *args, PyObject *kwds)
+{
+    PyObject *obj_x, *obj_bins, *obj_edges, *weights = Py_None;
+    PyArrayObject *arr_x = NULL, *arr_bins = NULL, *arr_edges = NULL,
+                  *arr_w = NULL, *hist = NULL;
+    npy_intp len, ndim, *bins;
+    double min[NPY_MAXDIMS], max[NPY_MAXDIMS], norm[NPY_MAXDIMS];
+
+    static char *kwlist[] = {"x", "nbins", "edges", "weights", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "OOO|O:_histogramdd_uniform",
+            kwlist, &obj_x, &obj_bins, &obj_edges, &weights)) {
+        goto fail;
+    }
+
+    /*
+     * Get data (x) which is of shape (len, ndim)
+     * where len is the number of samples
+     * and ndim is the number of dimensions
+     */
+    arr_x = (PyArrayObject *)PyArray_ContiguousFromAny(obj_x, NPY_DOUBLE, 2, 2);
+    if (arr_x == NULL) {
+        goto fail;
+    }
+    len = PyArray_DIM(arr_x, 0);
+    ndim = PyArray_DIM(arr_x, 1);
+
+    /*
+     * Get the number of bins for each array since
+     * the edges are passed in as a flat array
+     */
+    arr_bins = (PyArrayObject *)PyArray_ContiguousFromAny(obj_bins, NPY_INTP, 1, 1);
+    if (arr_bins == NULL) {
+        goto fail;
+    }
+    if (PyArray_DIM(arr_bins, 0) != ndim) {
+        PyErr_SetString(PyExc_ValueError,
+                        "Number of bins array must be same length as array");
+        goto fail;
+    }
+    bins = (npy_intp *)PyArray_DATA(arr_bins);
+
+    /* Handle empty array */
+    if (len == 0) {
+        hist = (PyArrayObject *)PyArray_ZEROS(ndim, bins, NPY_INTP, 0);
+        if (hist == NULL) {
+            goto fail;
+        }
+        Py_DECREF(arr_bins);
+        Py_DECREF(arr_x);
+        return (PyObject *)hist;
+    }
+
+    /*
+     * Get bin edges which is a flat array of length sum(nbins)
+     */
+    arr_edges = (PyArrayObject *)PyArray_ContiguousFromAny(obj_edges, NPY_DOUBLE, 1, 1);
+    if (arr_edges == NULL) {
+        goto fail;
+    }
+    {
+        npy_intp nedges_total = 0, i;
+        for (i = 0; i < ndim; i++) {
+            nedges_total += bins[i] + 1;
+        }
+        if (PyArray_SIZE(arr_edges) != nedges_total) {
+            PyErr_SetString(PyExc_ValueError,
+                            "Length of edges must be equal to the total number of edges");
+            goto fail;
+        }
+    }
+
+    {
```

Comment:
If they're awkward to pull into functions, it'd be good if each of these blocks could at least get a comment summarizing what it does.